### PR TITLE
[fix] XPU crash due to circle call of nonzero() and index_select() 

### PR DESCRIPTION
**Pull request by xin3he**
## Description

XPU cannot run with MOE model.

## Type of Change

Bug fix

## Related Issues

Fixes or relates to #1830

## Checklist Before Submitting

- [x] My code has been tested locally.
- [ ] Documentation has been updated as needed.
- [ ] New or updated tests are included where applicable.
- [ ] The CUDA CI has passed. You can trigger it by commenting `/azp run Unit-Test-CUDA-AutoRound`.

### CODE REVIEW

File: auto_round/modeling/fused_moe/utils.py [hunk 1]
```diff
@@ -39,6 +39,78 @@ def is_linearized_layout(original: torch.nn.Module) -> bool:
...
+        top_k_index: (num_tokens, top_k) selected expert indices.
+        top_k_weights: (num_tokens, top_k) routing weights.
+        experts: Indexable collection of per-expert callables (e.g. nn.ModuleList),
+            each taking (num_samples, hidden_dim) and returning (num_samples, hidden_dim).
+        num_experts: Total number of experts.
+
+    Returns:
+        final_hidden_states: (num_tokens, hidden_dim) output tensor.
+    """
+    hidden_dim = hidden_states.size(-1)
```
Comment (wenhuach21) [inline auto_round/modeling/fused_moe/utils.py:73]: is there any drawback in this implementation. If there is ,we should keeps two implementation and dispatch  them automatically

File: auto_round/modeling/fused_moe/moe_experts_interface.py [hunk 2]
```diff
@@ -225,17 +225,35 @@ def linear_loop_experts_forward(
...
-        if not mask.any():
+    # Group token-expert pairs by expert using a single sort, then run each
+    # expert on a contiguous *static* slice, instead of doing a per-expert
+    # nonzero()/boolean-mask lookup inside the Python loop.
+    #
+    # Why: repeatedly issuing dynamic-shape gather kernels (nonzero(),
+    # index_select() or boolean-mask indexing) once per expert in a loop
+    # reliably triggers a driver-level bug on some XPU builds - either a
+    # "vectorized gather kernel index out of bounds" device-side assertion or
+    # a hard UR_RESULT_ERROR_DEVICE_LOST - even when every index is valid.
```
Comment (yiliu30) [inline auto_round/modeling/fused_moe/moe_experts_interface.py:237]: Nice catch. Do we have issue or JIRA track this issue?

File: auto_round/modeling/fused_moe/utils.py (hunk 1 above)
Comment (xin3he) [reply auto_round/modeling/fused_moe/utils.py:73]: In my mind, no drawback but optimization.

File: auto_round/modeling/fused_moe/moe_experts_interface.py (hunk 2 above)
Comment (xin3he) [reply auto_round/modeling/fused_moe/moe_experts_interface.py:237]: https://github.com/intel/torch-xpu-ops/issues/4778

**Review by wenhuach21 — APPROVED** [review]
please make sure the accuracy is fine
